presto-bigquery/src/main/java/com/facebook/presto/plugin/bigquery/BigQueryMetadata.java

```diff
@@ -0,0 +1,242 @@
+        return bigQueryClient.getTable(TableId.of(projectId, tableName.getSchemaName(), tableName.getTableName()));
+    }
+
+    public ConnectorTableMetadata getTableMetadata(ConnectorSession session, SchemaTableName schemaTableName)
+    {
+        ConnectorTableHandle table = getTableHandle(session, schemaTableName);
+        if (table == null) {
+            throw new TableNotFoundException(schemaTableName);
+        }
+        return getTableMetadata(session, table);
+    }
+
+    @Override
+    public ConnectorTableMetadata getTableMetadata(ConnectorSession session, ConnectorTableHandle tableHandle)
+    {
+        log.debug("getTableMetadata(session=%s, tableHandle=%s)", session, tableHandle);
+        TableInfo table = bigQueryClient.getTable(((BigQueryTableHandle) tableHandle).getTableId());
+        SchemaTableName schemaTableName = new SchemaTableName(table.getTableId().getDataset(), table.getTableId().getTable());
+        Schema schema = table.getDefinition().getSchema();
+        List<ColumnMetadata> columns = schema == null ?
+                ImmutableList.of() :
+                schema.getFields().stream()
+                        .map(Conversions::toColumnMetadata)
+                        .collect(toImmutableList());
+        return new ConnectorTableMetadata(schemaTableName, columns);
+    }
+
+    @Override
+    public Map<String, ColumnHandle> getColumnHandles(ConnectorSession session, ConnectorTableHandle tableHandle)
+    {
+        log.debug("getColumnHandles(session=%s, tableHandle=%s)", session, tableHandle);
+        TableInfo table = bigQueryClient.getTable(((BigQueryTableHandle) tableHandle).getTableId());
+        Schema schema = table.getDefinition().getSchema();
+        return schema == null ?
+                ImmutableMap.of() :
+                schema.getFields().stream().collect(toMap(Field::getName, Conversions::toColumnHandle));
+    }
+
+    @Override
+    public ColumnMetadata getColumnMetadata(
+            ConnectorSession session,
+            ConnectorTableHandle tableHandle,
+            ColumnHandle columnHandle)
+    {
+        log.debug("getColumnMetadata(session=%s, tableHandle=%s, columnHandle=%s)", session, columnHandle, columnHandle);
+        return ((BigQueryColumnHandle) columnHandle).getColumnMetadata();
+    }
+
+    @Override
+    public Map<SchemaTableName, List<ColumnMetadata>> listTableColumns(ConnectorSession session, SchemaTablePrefix prefix)
+    {
+        log.debug("listTableColumns(session=%s, prefix=%s)", session, prefix);
+        requireNonNull(prefix, "prefix is null");
+        ImmutableMap.Builder<SchemaTableName, List<ColumnMetadata>> columns = ImmutableMap.builder();
+        for (SchemaTableName tableName : listTables(session, prefix)) {
+            try {
+                columns.put(tableName, getTableMetadata(session, tableName).getColumns());
+            }
+            catch (NotFoundException e) {
+                // table disappeared during listing operation
```

Comment:
throw exception?

Reply:
no actually it should be allowed that table disappeared during listing operation, transcation should finish instead of failing

Reply:
I do not think we prefer fail silently. Is there bigquery docs, showing it is OK for table disappear during list operation?

Reply:
no I think the purpose was no to interrupt the operation but it'll be the end of the operation so will throw a new BigQueryException here